django/control/api/views.py

```diff
@@ -21,25 +27,147 @@
+
+    def post(self, request):
+        # TODO check errors when writing to redis?
+        serializer = CreateBatchSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        data = serializer.validated_data
+
+        resource_ids = [resource.get("resource_id") for resource in data["resources"]]
+
+        authorization_header = request.META.get("HTTP_AUTHORIZATION")
+
+        batch_id = str(uuid.uuid4())
+        batch_timestamp = datetime.now().strftime("%Y-%m-%dT%H:%M:%S")
+
+        # Add batch info to redis
+        batch_counter_redis = redis.Redis(
+            host=settings.REDIS_COUNTER_HOST,
+            port=settings.REDIS_COUNTER_PORT,
+            db=settings.REDIS_COUNTER_DB,
+        )
+        batch_counter_redis.hset("batch", batch_id, batch_timestamp)
+        batch_counter_redis.sadd(f"batch:{batch_id}:resources", *resource_ids)
+
+        # Create kafka topics for batch
+        # TODO num parts, etc. in env
+        new_topics = [
+            NewTopic(f"batch.{batch_id}", 1, 1),
+            NewTopic(f"extract.{batch_id}", 1, 1),
+            NewTopic(f"transform.{batch_id}", 1, 1),
+            NewTopic(f"load.{batch_id}", 1, 1),
+        ]
+        admin_client = AdminClient({"bootstrap.servers": settings.KAFKA_BOOTSTRAP_SERVERS})
+        admin_client.create_topics(new_topics)
+
+        # Fetch mapping
+        mappings_redis = redis.Redis(
+            host=settings.REDIS_MAPPINGS_HOST, port=settings.REDIS_MAPPINGS_PORT, db=settings.REDIS_MAPPINGS_DB
+        )
+
+        for resource_id in resource_ids:
+            resource_mapping = fetch_resource_mapping(resource_id, authorization_header)
+            mappings_redis.set(f"{batch_id}:{resource_id}", json.dumps(resource_mapping))
+
+        # Delete documents from previous batch
+        for resource in data["resources"]:
+            resource_id = resource.get("resource_id")
+            resource_type = resource.get("resource_type")
+            logger.debug(
+                {
+                    "message": f"Deleting all documents of type {resource_type} for given resource",
+                    "resource_id": resource_id,
+                },
+            )
+
+            fhirstore = get_fhirstore()
+            try:
+                fhirstore.delete(resource_type, resource_id=resource_id)
+            except NotFoundError:
+                logger.debug(
+                    {"message": f"No documents for resource {resource_id} were found", "resource_id": resource_id},
+                )
+
+        # Send event to the extractor
+        producer = Producer(broker=settings.KAFKA_BOOTSTRAP_SERVERS)
+        for resource_id in resource_ids:
+            event = {"batch_id": batch_id, "resource_id": resource_id}
+            producer.produce_event(topic=f"batch.{batch_id}", event=event)
+
+        return Response({"id": batch_id, "timestamp": batch_timestamp}, status=status.HTTP_200_OK)
+
+
+class DeleteBatchEndpoint(views.APIView):
```

Comment:
is it required to use a model with a `ViewSet` ? I though we could use a more "custom" implementation than directly mapping routes to database objects.
https://medium.com/django-rest-framework/django-rest-framework-viewset-when-you-don-t-have-a-model-335a0490ba6f

Reply:
You're right ! :) (https://www.django-rest-framework.org/api-guide/viewsets/#viewsets)
I'm so used to use `viewsets.ModelViewSet` :sweat_smile:

I still think using a model (postgres) and `viewsets.ModelViewSet` would be much more straightforward. We can discuss it later. In any case, I don't expect this to be in the PR. Ok for the simple `ViewSet`

Reply:
I also think it may be a good idea to store the list of batches in Postgres along with their resource ids. The counter (very write intensive) and the mapping should stay in Redis though imo.